crates/aggregate_builder/Readme.md

```diff
@@ -28,15 +28,15 @@ Current syntax looks something like like
     // parallel-safety marker if desireable
     const PARALLEL_SAFE: bool = true;
 
-    fn serialize(state: Inner<State>) -> bytea {
+    fn serialize(state: &State) -> bytea {
         // serialize function body goes here
     }
 
     fn deserialize(bytes: bytea) -> Inner<State> {
         // deserialize function body goes here
     }
 
-    fn combine(state1: Option<Inner<State>>, state2: Option<Inner<State>>) -> Option<Inner<State>> {
+    fn combine(state1: Option<&State>, state2: Option<&State>) -> Option<Inner<State>> {
```

Comment:
Given that this return is the only surviving `Inner`, I'd try to wrap this to hide that layer entirely.

Reply:
🤔 I didn't believe I could last time I checked, but maybe it'll work now...

Reply:
ah, this is `fn combine()` not `fn transition()` I think I actually can hide it here. That said, we're also using it in `fn combine()` and `fn transition()` and I'm not sure if we should do one without the others.

---

The issue with `fn transition()`: in _theory_ it would be nice to give `fn transition()` a signature like
```rust
fn transition(state: Option<State>) -> Option<State>
```
unfortunately, this runs into weird issues around `panic!()`s: desugared, the call procedure for the inner function would look something like
```rust
fn transition_like(state: Option<Inner<State>>) -> Option<Inner<State>> {
    match state {
        None => inner(None).map(|s| s.into()),
        Some(mut state) => {
            let output = inner(Some(*state));
            match output {
                None => None,
                Some(new) => {
                    *state = new;
                    Some(state)
                }
            }
        }
    }
}

fn inner(state: Option<String>) -> Option<String> { /* elided */ }
```
now this _almost_ works, and the equivalent code is safe for box ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82e3747cd9f2c89cdbd289481332f28f)) the issue we run into is that we want `Inner<State>` to be recursively freed when the aggregate memory context is discarded. If this occurs due to an error in `inner(Some(state))` we need some way to communicate to `Inner`'s `Drop` impl that `state` has been moved out and it can only drop the pointer, not its contents. Right now I only know of one way to do this: store a dynamic drop flag. That would make the transition function desugar to something more like:

```rust
fn transition_like(state: Option<Inner<Option<String>>>) -> Option<Inner<Option<String>>> {
    match state {
        None => inner(None).map(|s| s.into()),
        Some(mut state) => {
            let inner_state = state.take();
            let output = inner(Some(inner_state));
            match output {
                None => None,
                Some(new) => {
                    *state = Some(new);
                    Some(state)
                }
            }
        }
    }
}
```
Is it worth storing the extra flag so we don't need to think about `Inner<>` as much? I'm not really sure. Realistically, I don't think the performance would change enough to matter, and the complexity isn't _too_ bad, but then, I don't see the complexity being saved as all that much either...

Reply:
Note that there _may_ be a future compatibility hazard with window functions; they may require a special input to get the tuplestore out of. I don't think this really matters here though.